e3sm_diags/driver/utils/dataset_xr.py

```diff
@@ -0,0 +1,999 @@
+
+        return filepath
+
+    def _get_climo_filepath_with_params(self) -> str | None:
+        """Get the climatology filepath using parameters.
+
+        Returns
+        -------
+        str | None
+            The filepath using the `ref_file` or `test_file`  parameter if they
+            are set.
+        """
+        filepath = None
+
+        if self.data_type == "ref":
+            if self.parameter.ref_file != "":
+                filepath = os.path.join(self.root_path, self.parameter.ref_file)
+
+        elif self.data_type == "test":
+            if hasattr(self.parameter, "test_file"):
+                filepath = os.path.join(self.root_path, self.parameter.test_file)
+
+        return filepath
+
+    def _find_climo_filepath(self, filename: str, season: str) -> str | None:
+        """Find the climatology filepath for the variable.
+
+        Parameters
+        ----------
+        filename : str
+            The filename for the climatology variable.
+        season : str
+            The season for climatology.
+
+        Returns
+        -------
+        str | None
+            The filepath for the climatology variable.
+        """
+        # First attempt: try to find the climatology file based on season.
+        # Example: {path}/{filename}_{season}.nc
+        filepath = self._find_climo_filepath_with_season(
+            self.root_path, filename, season
+        )
+
+        # Second attempt: try looking for the file nested in a folder, based on
+        # the test_name.
+        # Example: {path}/{filename}/{filename}_{season}.nc
+        #           data_path/some_file/some_file_ANN.nc
+        if filepath is None:
+            nested_root_path = os.path.join(self.root_path, filename)
+
+            if os.path.exists(nested_root_path):
+                filepath = self._find_climo_filepath_with_season(
+                    nested_root_path, filename, season
+                )
+
+        return filepath
+
+    def _find_climo_filepath_with_season(
+        self, root_path: str, filename: str, season: str
+    ) -> str | None:
+        """Find climatology filepath with a root path, filename, and season.
+
+        Parameters
+        ----------
+        root_path : str
+            The root path containing `.nc` files. The `.nc` files can be nested
+            in sub-directories within the root path.
+        filename : str
+            The filename for the climatology variable.
+        season : str
+            The season for climatology.
+
+        Returns
+        -------
+        str | None
+            The climatology filepath based on season, if it exists.
+        """
+        files_in_dir = sorted(os.listdir(root_path))
+
+        # If the filename is followed by _<SEASON>.
+        for file in files_in_dir:
+            if file.startswith(filename + "_" + season):
+                return os.path.join(root_path, file)
+
+        # For model only data, the <SEASON> string can by anywhere in the
```

Comment:
It seems like this code block can preserve the feature that when `ref_name` is given an empty string, the climo files can be still picked up based on the season string in the file? as mentioned [here](https://github.com/E3SM-Project/e3sm_diags/issues/736#issuecomment-1751458904).

Reply:
Yeah I ported the code from here:
https://github.com/E3SM-Project/e3sm_diags/blob/a2d00eb114cf505a4ecf4dd429bb8710778cc087/e3sm_diags/driver/utils/dataset.py#L323-L326

but moved the `if season in ["ANN",...` outside of the loop to only check `season` once.